src/cs/Ssh/SshSession.cs

```diff
@@ -899,25 +904,15 @@ await protocol.SendMessageAsync(message, cancellationSource.Token)
 	private async Task HandleMessageAsync(
 		KeyExchangeMessage message, CancellationToken cancellation)
 	{
-		if (this.kexService == null)
+		if (this.kexService != null)
 		{
-			var protocol = Protocol;
-			if (message is KeyExchangeInitMessage && protocol != null)
-			{
-				// This side didn't configure security, but the other side still wants to negotiate.
-				// Start the KEX sequence just to try to negotiate 'none'.
-				this.kexService = ActivateService<KeyExchangeService>();
-				protocol.KeyExchangeService = this.kexService;
-				await protocol.ConsiderReExchangeAsync(initial: true, cancellation).ConfigureAwait(false);
-			}
-			else
-			{
-				await CloseAsync(SshDisconnectReason.KeyExchangeFailed).ConfigureAwait(false);
-				return;
-			}
+			await this.kexService.HandleMessageAsync(message, cancellation).ConfigureAwait(false);
+		}
+		else if (!(message is KeyExchangeInitMessage initMessage && initMessage.AllowsNone))
+		{
+			// The other side required some security, but it's not configured here.
```

Comment:
Should we add a trace too?

Reply:
The outgoing disconnect message will be traced already, however I see that it doesn't include the reason code. I'll add a `ToString()` override for that.

Reply:
Updated.